loki/visitors/pretty_visualize.py

```diff
@@ -0,0 +1,349 @@
+        """
+        Dispatch routine to add nodes utilizing expression tree stringifier,
+        mark parent and stop.
+
+        Returns
+        -------
+        list[tuple[dict[str,str], dict[str,str]]]]
+            An extended list of tuples of a node and potentially a edge information or list of nothing.
+        """
+        if kwargs.get("show_expressions"):
+            content = self.symgen(o)
+            parent = kwargs.get("parent")
+            return self.__add_node(o, label=content, parent=parent, shape="box")
+        return [None]
+
+    def visit_tuple(self, o, **kwargs):
+        """
+        Recurse for each item in the tuple.
+        """
+        return self.visit_all(o, **kwargs)
+
+    visit_list = visit_tuple
+
+    def visit_Conditional(self, o, **kwargs):
+        """
+        Add a :any:`Conditional`, mark parent and visit first body then else body.
+
+        Returns
+        -------
+        list[tuple[dict[str,str], dict[str,str]]]]
+            An extended list of tuples of a node and potentially a edge information
+        """
+        parent = kwargs.get("parent")
+        label = self.symgen(o.condition)
+        node_edge_info = self.__add_node(o, label=label, parent=parent, shape="diamond")
+        kwargs["parent"] = o
+        node_edge_info.extend(self.visit_all(o.body, **kwargs))
+
+        if o.else_body:
+            node_edge_info.extend(self.visit_all(o.else_body, **kwargs))
+        return node_edge_info
+
+
+def pretty_visualize(ir, **kwargs):
+    """
+    Pretty-print the given IR using :class:`Visualizer`.
+
+    Parameters
+    ----------
+    ir : :any:`Node`
+        The IR node starting from which to produce the tree
+    kwargs["filename"] : str, optional
+        The location, name and format of the output graph, if none is given the graph
+        object is returned. In jupyter-notebooks this imideatly plots it, in terminal
+        it just shows a text representation.
+    kwargs["show_comments"] : bool, optional, default: True
+        Whether to show comments in the output
+    kwargs["show_expressions"] : bool, optional, default: False
+        Whether to further expand expressions in the output
+    """
+    filename = kwargs.get("filename")
+    if filename is None:
+        log = "[Loki::Graph Visualization] Creating graph visualization in {:.2f}s"
+    else:
+        log = f"[Loki::Graph Visualization] Visualized to {filename}" + " in {:.2f}s"
+
+    with Timer(text=log):
+        visualizer = Visualizer()
+        node_edge_info = [
+            item for item in visualizer.visit(ir, **kwargs) if item is not None
+        ]
+
+        graph = Digraph()
+        graph.attr(rankdir="LR")
+        for node_info, edge_info in node_edge_info:
+            if node_info:
+                graph.node(**node_info)
+            if edge_info:
+                graph.edge(**edge_info)
+        if filename:
+            graph.render(**kwargs)
+        else:
+            return graph
```

Comment:
I think we should either always or never return the graph, conditionally switching off return values is considered bad practice (and will likely be flagged by the linter)

Reply:
I agree, since graph is always returned rendering of graph is no longer reasonable, therefore filename is removed c72fe87359edd81a3f233e85478edf849f7b2c33